File: SimulationPython/Simulation/Sphere.py
import numpy as np
import random

from SimulationPython.Simulation.AbstractCompartment import AbstractCompartment

class Sphere(AbstractCompartment):
    def __init__(self, x, y, z, T2, diffusivity, permeability, radius):
        super(Sphere, self).__init__(x, y, z, T2, diffusivity)
        self.permeability = permeability
        self.radius = radius
        self.parentComp = None

    def setParentComp(self, compartment):
        self.parentComp = compartment

    def findIntersection(self, ray, maxDistance):
        #ray direction vector must be normalized!
        if np.linalg.norm(ray[0] - self.pos) - self.radius <= maxDistance:
            L = self.pos - ray[0]
            b = -2*np.dot(ray[1], L)
            c = np.linalg.norm(L)**2 - self.radius**2
            delta = b**2 - 4*c
            if delta > 0:
                t = (-b - delta**0.5) / 2
                if t > 0:
                    return t
                else:
                    t = (-b + delta**0.5) / 2
                    if t > 0:
                        return t
            elif delta == 0:
                t = -b/2
                if t > 0:
                    return t
        return np.inf

    def collide(self, particle, oldPos, intersection, sim):
        if self.contains(oldPos):
            #we are leaving this compartment
            otherSideComp = self.parentComp
        else:
            #we are entering this compartment
            otherSideComp = self
        transmissionProba = 4*self.permeability/particle.getSpeed()
        assert(transmissionProba <= 1)

        if random.random() > transmissionProba:
            #deflection
            normal = (intersection - self.pos)
            normal = normal/np.linalg.norm(normal)
            particle.setVelocity(particle.getVelocity() - 2*normal*np.dot(normal, particle.getVelocity()))
        else:
            particle.changeCompartment(otherSideComp, sim.getTimeStep())

    def contains(self, pos):
        return np.linalg.norm(pos - self.pos) < self.radius

    def plot(self, ax):
        ############################################
        u, v = np.mgrid[0:2*np.pi:20j, 0:np.pi:10j]
        x = self.pos[0] + self.radius*np.cos(u)*np.sin(v)
        y = self.pos[1] + self.radius*np.sin(u)*np.sin(v)
        z = self.pos[2] + self.radius*np.cos(v)
        ax.plot_wireframe(x, y, z, color="black")
        #############################################

File: SimulationCpp/SimulationCpp/src/Sphere.cpp
#include "Sphere.h"
#include "Simulation.h"
#include "Util.h"

#include <iostream>

Sphere::Sphere(double x, double y, double z, double T2, double diffusivity, double permeability, double radius)
	:AbstractCompartment{x, y, z, T2, diffusivity},
	m_permeability{ permeability },
	m_radius{ radius }
{
}

bool Sphere::contains(Eigen::Vector3d pos) {
	return (pos - m_pos).norm() < m_radius;
}

double Sphere::findIntersection(std::array<Eigen::Vector3d, 2> ray, double maxDistance) {
	//ray direction vector must be normalized!
	Eigen::Vector3d L = m_pos - ray[0];
	if (L.norm() - m_radius <= maxDistance) {
		double b = -2 * L.dot(ray[1]);
		double c = L.squaredNorm() - m_radius*m_radius;
		double delta = b * b - 4 * c;	
		if (delta > 0) {
			double t = (-b - sqrt(delta)) / 2;
			if (t > 0) {
				return t;
			}
			else {
				t = (-b + sqrt(delta)) / 2;
				if (t > 0) {
					return t;
				}
			}
		}
		else if (delta == 0) {
			double t = -b / 2;
			if (t > 0) {
				return t;
			}
		}
	}
	//no intersection
	return std::numeric_limits<double>::infinity();
}

bool Sphere::collide(std::shared_ptr<AbstractParticle3D> particle, Eigen::Vector3d oldPos, Eigen::Vector3d intersection, double timeStep) {
	double transmissionProba = 4 * m_permeability / particle->getSpeed();
	if (transmissionProba > 1) {
		std::cout << "Error: greater than one transmission probability!" << std::endl;
	}
	assert(transmissionProba <= 1);

	if (Util::getRandomUniform(0, 1) >= transmissionProba) {
		//deflection
		Eigen::Vector3d normal = (intersection - m_pos).normalized();
		particle->setVelocity(particle->getVelocity() - 2 * normal*normal.dot(particle->getVelocity()));
	} 
	else {
		//transmission
		if (contains(oldPos)) {
			//we are leaving this compartment and going to parent
			particle->changeCompartment(m_parentComp, timeStep);
		}
		else {
			//we are entering this compartment
			particle->changeCompartment(std::make_shared<Sphere>(*this), timeStep);
		}
	}
	return false;
}
